multiuser/machine-auth/che-multiuser-machine-authentication/src/main/java/org/eclipse/che/multiuser/machine/authentication/server/signature/SignatureKeyManager.java

```diff
@@ -134,13 +138,15 @@ public void removeKeyPair(String workspaceId) {
   /** Loads signature key pair if no existing keys found then stores a newly generated key pair. */
   @PostConstruct
   @VisibleForTesting
-  KeyPair loadKeyPair(String workspaceId) throws ServerException, ConflictException {
+  KeyPair loadKeyPair(String workspaceId)
+      throws ServerException, ConflictException, NotFoundException {
     try {
       return toJavaKeyPair(signatureKeyDao.get(workspaceId));
     } catch (NotFoundException nfe) {
       try {
+        workspaceManager.getWorkspace(workspaceId); // Make sure workspace still exists
```

Comment:
I think it doesn't guarantee data integrity.
Maybe it would be better to catch `IntegrityConstraintViolationException` in DAO and remap it into a meaningful exception?